### PR TITLE
Move `Enumerable` `to_json` outside of loop of other modules to fix #20775

### LINKED ISSUE #20775: ActiveSupport 4.2 - `JSON.dump` broken for Structs
Hi everyone,

Today I've noticed that ActiveSupport 4.2 breaks `JSON.dump` ("stack level too deep") for the arrays/hashes containing structs.

Example:
```ruby
irb(main):001:0> require 'json'
=> true
irb(main):002:0> obj = Struct.new(:key).new
=> #<struct key=nil>
irb(main):003:0> JSON.dump([obj])
=> "[\"#<struct key=nil>\"]"
irb(main):004:0> require 'active_support/json'
=> true
irb(main):005:0> JSON.dump([obj])
SystemStackError: stack level too deep
```

Stack trace shows that this line `activesupport-4.2.0/lib/active_support/core_ext/object/json.rb:34:in `to_json_with_active_support_encoder'` is called over and over.

I tried to figure out why is it failing, but I couldn't find the root of the problem.
If you have any hint on how I can provide more useful information, let me know.

This issue appears in ActiveSupport 4.2.x (I've checked 4.2.0 and 4.2.3), but it does not appear in 4.1.x

### PR DESCRIPTION
**Pull request by codeodor**
<s>I cherry picked d5bddc1 from the master branch to fix https://github.com/rails/rails/issues/20775</s>

Cherry picking was the wrong solution. See the commit message instead.

### CODE REVIEW

**Comment by matthewd** [comment]
This seems very unlikely to work on 1.9.

**Comment by chancancode** [comment]
Yeah, that's the problem, 4.2 needs to support 1.9, so we'll need to find a better fix

**Comment by codeodor** [comment]
Ok, I'll work on that.

**Comment by codeodor** [comment]
Ok, I did not run all the tests, but the one relevant to #20775 passes on Ruby 2.2.2 and 1.9.3.

**Comment by codeodor** [comment]
grrr never mind, I see what you mean. Lots of errors in the test suite now.

**Comment by codeodor** [comment]
Ok, should be good after that latest commit.

**Comment by codeodor** [comment]
Just changed the commit message and added a comment to the code, so future programmers will know why it is not done in the loop with the rest of the modules.

**Comment by chancancode** [comment]
I still have some doubts that this is "correct", I'll try to think harder about why I think it will break and report back. Meanwhile, can you add some tests? :P

**Comment by codeodor** [comment]
@chancancode Is there a way to force the Rails test suite to run a specific file in a newly loaded environment when running the whole suite?

If not, I do not see the value in a test, since this only occurs if we were to unload files other tests have already loaded. :(

**Comment by codeodor** [comment]
I say that after trying to write a test for the last 45 minutes which fails before the change while running the whole suite, and I have been unable to figure out how. It's not that I did not try :)

**Comment by chancancode** [comment]
Yes there is a way :smile: https://github.com/rails/rails/blob/master/activesupport/test/core_ext/object/json_cherry_pick_test.rb is probably the most analogous example.

**However**, if I am understanding it correctly, the problem is that we promised to not interfere with the output of `JSON.{dump,generate}(...)`  with or without Active Support being loaded. So we should just test this normally with AS loaded, no? (Otherwise we would just be testing JSON gem by itself which won't be very interesting.)

**Comment by codeodor** [comment]
Oh, fantastic. I will try that. Thanks!

**Comment by chancancode** [comment]
(Make sure you read the second half before you spend an hour+ on this, my hunch is that you won't actually need the first file, since *I think* we are just testing `assert_equal "...", JSON.generate(...)` :smile:)

**Comment by chancancode** [comment]
(Specifically: https://github.com/rails/rails/blob/master/activesupport/test/json/encoding_test.rb#L424-L445)

**Comment by codeodor** [comment]
So, when using `ActiveSupport::Testing::Isolation` I guess `rake test` skips over it, and the CI knows it will need to run them individually?  I ask because when I include it, `rake test` everything passes, but running the file individually, I get failures.

**Comment by codeodor** [comment]
And I suppose I see your point on "..we promised to not interfere with the output of..." 

As it is, this does turn a struct into what one might expect, a hash, while `JSON.dump` turns it to a String. I don't know why anyone would want a string, but I will see what I can do to get it to match the output when using JSON.dump with or without the core-ext applied.

**Comment by chancancode** [comment]
> So, when using ActiveSupport::Testing::Isolation I guess rake test skips over it, and the CI knows it will need to run them individually? I ask because when I include it, rake test everything passes, but running the file individually, I get failures.

Ah, I remember why! When you run the whole test suite I think the "isolated" environment might already have other dependencies loaded depending on the ordering of the test cases (it forks from the current process). On CI (and when not in a pull request) it runs `rake test:isolated` (followed by normal `rake test`) which run each file in a new environment.

>  I don't know why anyone would want a string, but I will see what I can do to get it to match the output when using JSON.dump with or without the core-ext applied.

It's less about the output specifically, but more about that we don't interfere with how JSON gem work. Unfortunately we both use the `to_json` method but with subtlely different meaning/behavior, and traditionally we used to "fight" over it and caused a lot of problems/surprises, so in 4.1+ we try very hard to co-exist – namely when we know that we are called by `JSON.{dump,generate}` (by looking at the argument), we let JSON gem do its thing. (Which is part of why I suspect that the fix is incorrect, because it doesn't consider the argument.)

**Comment by codeodor** [comment]
Totally understand! Thanks for the explanation.

The latest rebased commit now includes tests that fail when run in isolation without the fixed code, and those tests ensure we are not messing with JSON gem's output.

**Comment by chancancode** [comment]
Interesting fix, thanks for working on this! I'll check after work today :grin:

**Comment by codeodor** [comment]
Yeah, I think what happened is `Struct` got it from `Object`, then again from `Enumerable` but by that time `to_json` was already pointing at itself, to it pointed to itself again, causing the recursion.

**Comment by chancancode** [comment]
Ahhh I figured out why this works! Our implementation (before and after this patch) technically changes behavior for `JSON.{dump,generate}` on `BasicObject` that includes `Enumerable` and defines `respond_to?`......... :stuck_out_tongue: 

....but I'll pretend no one cares about that until someone complains :trollface:

File: activesupport/lib/active_support/core_ext/object/json.rb [hunk 1]
```diff
@@ -26,7 +26,11 @@
 # bypassed completely. This means that as_json won't be invoked and the JSON gem will simply
 # ignore any options it does not natively understand. This also means that ::JSON.{generate,dump}
 # should give exactly the same results with or without active support.
-[Object, Array, FalseClass, Float, Hash, Integer, NilClass, String, TrueClass, Enumerable].each do |klass|
+#
+# Enumerable needs to come before the others because otherwise infinite recursion
+# results in the call to  `to_json_without_active_support_encoder`
+
+[Enumerable, Object, Array, FalseClass, Float, Hash, Integer, NilClass, String, TrueClass].each do |klass|
```
Comment (prathamesh-sonpatki) [inline activesupport/lib/active_support/core_ext/object/json.rb]: Yeah :smile: I was just thinking about this last night :smile:

**Comment by chancancode** [comment]
Thanks so much for fixing this! Merged manually in 49c613463b758a520a6162e702acc1158fc210ca (credited both of you) with tests added in d6142c74d1f9decc2687aa388899a76a03ab0e49 :smile:

**Comment by prathamesh-sonpatki** [comment]
Thanks :smile:

**Comment by codeodor** [comment]
:+1:  Thanks!

**Pull request closed without merging**